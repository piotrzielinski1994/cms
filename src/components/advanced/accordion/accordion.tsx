'use client';

import { HtmlProps } from '@/utils/html/html.types';
import { cn } from '@/utils/tailwind';
import { BoolMap } from '@/utils/types';
import { ChevronDown } from 'lucide-react';
import { forwardRef, ReactNode, useEffect, useId, useRef, useState } from 'react';

type AccordionProps = HtmlProps<'div'> & {
  items: {
    heading: ReactNode;
    content: ReactNode;
  }[];
  activeItemIndex?: number;
};

const styles = {
  root: cn('grid gap-2', 'cms-accordion'),
  item: cn('bg-components-accordion text-components-accordion-foreground', 'group'),
  itemHeader: ({ isActive }: BoolMap<'isActive'>) =>
    cn(
      'w-full px-4 py-2 md:px-6 md:py-4',
      'flex justify-between items-center gap-2 text-lg cursor-pointer select-none',
      { 'outline outline-2 outline-offset-2 outline-blue-500': isActive },
      'tw-has-focus:tw-cms-outline',
    ),
  icon: ({ isActive }: BoolMap<'isActive'>) =>
    cn('h-[1em] w-[1em] transition-transform', { 'rotate-180': isActive }),
  radio: 'sr-only',
  contentWrapper: 'overflow-hidden transition-[max-height] duration-300 ease-in-out',
  content: 'px-4 pb-4 md:px-6',
};

const Component = ({ items, activeItemIndex, ...rest }: AccordionProps) => {
  const [activeIndex, setActiveIndex] = useState(activeItemIndex);
  const contentRefs = useRef<HTMLDivElement[]>([]);
  const id = useId();

  useEffect(() => {
    contentRefs.current.forEach((content, index) => {
      if (!content) return;
      content.style.maxHeight = activeIndex === index ? content.scrollHeight + 'px' : '0';
    });
  }, [activeIndex]);

  return (
    <Root {...rest}>
      {items.map(({ heading, content }, index) => {
        const isActive = activeIndex === index;
        return (
          <Item key={String(heading)}>
            <ItemHeader
              isActive={isActive}
              htmlFor={`${id}__${index}__input`}
              id={`${id}__${index}__label`}
            >
              <Radio
                name={`${id}-accordion`}
                id={`${id}__${index}__input`}
                checked={isActive}
                onChange={() => setActiveIndex(isActive ? undefined : index)}
                onClick={() => setActiveIndex(isActive ? undefined : index)}
                onKeyDown={(e) => {
                  if (![' ', 'Enter'].includes(e.key)) return;
                  e.preventDefault();
                  setActiveIndex(activeIndex === index ? undefined : index);
                }}
              />
              <span>{heading}</span>
              <Icon isActive={isActive} />
            </ItemHeader>
            <ContentWrapper
              id={`${id}__${index}__content`}
              style={{
                maxHeight: isActive ? contentRefs.current[index]?.scrollHeight + 'px' : '0',
              }}
              ref={(el) => {
                if (!el) return;
                contentRefs.current[index] = el;
              }}
            >
              <Content aria-labelledby={`${id}__${index}__label`} aria-hidden={!isActive}>
                {content}
              </Content>
            </ContentWrapper>
          </Item>
        );
      })}
    </Root>
  );
};

const Root = ({ className, ...rest }: HtmlProps<'div'>) => {
  return <div role="radiogroup" {...rest} className={cn(styles.root, className)}></div>;
};

const Item = ({ className, ...rest }: HtmlProps<'div'>) => {
  return <div {...rest} className={cn(styles.item, className)}></div>;
};

const ItemHeader = ({ isActive, className, ...rest }: HtmlProps<'label'> & BoolMap<'isActive'>) => {
  return <label {...rest} className={cn(styles.itemHeader({ isActive }), className)}></label>;
};

const Radio = ({ className, ...rest }: HtmlProps<'input'>) => {
  return <input type="radio" {...rest} className={cn(styles.radio, className)} />;
};

const Icon = ({ isActive, className, ...rest }: HtmlProps<'svg'> & BoolMap<'isActive'>) => {
  return <ChevronDown {...rest} className={cn(styles.icon({ isActive }), className)} />;
};

const ContentWrapper = forwardRef<HTMLDivElement, HtmlProps<'div'>>((props, ref) => {
  const { className, ...rest } = props;
  return <div ref={ref} {...rest} className={cn(styles.contentWrapper, className)}></div>;
});

const Content = ({ className, ...rest }: HtmlProps<'div'>) => {
  return <div role="region" {...rest} className={cn(styles.content, className)}></div>;
};

const Accordion = Object.assign(Component, {
  Root,
  Item,
  ItemHeader,
  Radio,
  Icon,
  ContentWrapper,
  Content,
});

export { Accordion, styles };
