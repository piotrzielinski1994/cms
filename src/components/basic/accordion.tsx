'use client';

import { cn } from '@/utils/tailwind';
import { ChevronDown } from 'lucide-react';
import React, { useEffect, useId, useRef, useState } from 'react';

type AccordionProps = {
  items: {
    heading: React.ReactNode;
    content: React.ReactNode;
  }[];
};

const Accordion = ({ items }: AccordionProps) => {
  const id = useId();
  const [openIndex, setOpenIndex] = useState<number | null>(null);
  const contentRefs = useRef<HTMLDivElement[]>([]);

  useEffect(() => {
    if (openIndex === null) return;
    const content = contentRefs.current[openIndex];
    if (content) content.style.maxHeight = content.scrollHeight + 'px';
  }, [openIndex]);

  return (
    <div className="grid gap-2">
      {items.map(({ heading, content }, index) => {
        const isOpen = openIndex === index;
        return (
          <div
            key={String(heading)}
            className="bg-components-accordion text-components-accordion-foreground"
          >
            <button
              type="button"
              className="w-full px-4 py-2 flex justify-between items-center gap-2 text-lg"
              onClick={() => setOpenIndex(openIndex === index ? null : index)}
              aria-controls={`${id}__${index}`}
              aria-expanded={isOpen}
            >
              <span>{heading}</span>
              <ChevronDown
                className={cn('h-[1em] w-[1em] transition-transform', {
                  'rotate-180': isOpen,
                })}
              />
            </button>
            <div
              id={`${id}__${index}`}
              className="overflow-hidden transition-[max-height] duration-300 ease-in-out"
              style={{ maxHeight: isOpen ? 'auto' : '0' }}
              ref={(el) => {
                if (!el) return;
                contentRefs.current[index] = el;
              }}
            >
              <div className="px-4 py-2">{content}</div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export { Accordion };
