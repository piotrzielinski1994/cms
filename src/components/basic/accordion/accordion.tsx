'use client';

import { cn } from '@/utils/tailwind';
import { ChevronDown } from 'lucide-react';
import React, { HTMLAttributes, useEffect, useId, useRef, useState } from 'react';

type AccordionProps = HTMLAttributes<HTMLDivElement> & {
  items: {
    heading: React.ReactNode;
    content: React.ReactNode;
  }[];
  activeItemIndex?: number;
};

const Accordion = ({ items, activeItemIndex, className, ...props }: AccordionProps) => {
  const [activeIndex, setActiveIndex] = useState(activeItemIndex);
  const contentRefs = useRef<HTMLDivElement[]>([]);
  const id = useId();

  useEffect(() => {
    contentRefs.current.forEach((content, index) => {
      if (!content) return;
      content.style.maxHeight = activeIndex === index ? content.scrollHeight + 'px' : '0';
    });
  }, [activeIndex]);

  return (
    <div {...props} className={cn('grid gap-2', 'cms-accordion', className)} role="radiogroup">
      {items.map(({ heading, content }, index) => {
        const isActive = activeIndex === index;
        return (
          <div
            key={String(heading)}
            className={cn('bg-components-accordion text-components-accordion-foreground', 'group')}
          >
            <label
              htmlFor={`${id}__${index}__input`}
              id={`${id}__${index}__label`}
              className={cn(
                'w-full px-4 py-2 flex justify-between items-center gap-2 text-lg cursor-pointer select-none',
                { 'outline outline-2 outline-offset-2 outline-blue-500': isActive },
                'focus-within:tw-cms-outline',
              )}
            >
              <input
                type="radio"
                name={`${id}-accordion`}
                id={`${id}__${index}__input`}
                checked={isActive}
                onChange={() => setActiveIndex(isActive ? undefined : index)}
                onClick={() => setActiveIndex(isActive ? undefined : index)}
                onKeyDown={(e) => {
                  if (![' ', 'Enter'].includes(e.key)) return;
                  e.preventDefault();
                  setActiveIndex(activeIndex === index ? undefined : index);
                }}
                className="sr-only"
                aria-checked={isActive}
              />
              <span>{heading}</span>
              <ChevronDown
                className={cn('h-[1em] w-[1em] transition-transform', { 'rotate-180': isActive })}
              />
            </label>
            <div
              id={`${id}__${index}__content`}
              className="overflow-hidden transition-[max-height] duration-300 ease-in-out"
              style={{
                maxHeight: isActive ? contentRefs.current[index]?.scrollHeight + 'px' : '0',
              }}
              ref={(el) => {
                if (!el) return;
                contentRefs.current[index] = el;
              }}
            >
              <div
                className="px-4 py-2"
                role="region"
                aria-labelledby={`${id}__${index}__label`}
                aria-hidden={!isActive}
              >
                {content}
              </div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export { Accordion };
