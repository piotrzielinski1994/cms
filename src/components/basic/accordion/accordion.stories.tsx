import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { Accordion as AccordionComponent } from './accordion';

type Args = ComponentProps<typeof AccordionComponent>;

const meta: Meta<Args> = {
  component: AccordionComponent,
  title: 'Components/Basic/Accordion',
  argTypes: {
    items: { control: 'object' },
    activeItemIndex: { control: 'number' },
  },
  args: {
    items: [
      { heading: DEFAULT_VALUE, content: DEFAULT_VALUE },
      { heading: DEFAULT_VALUE, content: DEFAULT_VALUE },
      { heading: DEFAULT_VALUE, content: DEFAULT_VALUE },
    ],
    activeItemIndex: 0,
  },
};

const Render = (args: Args) => {
  const t = useTranslations('storybook.basic.accordion');
  const props = {
    ...args,
    items: args.items.map((item, index) => ({
      heading: getFallback(item.heading, `${t('heading')} ${index + 1}`),
      content: getFallback(item.content, `${t('content')} ${index + 1}`),
    })),
  };
  return <AccordionComponent key={args.activeItemIndex} {...props} />;
};

const Accordion: StoryObj<typeof AccordionComponent> = { render: Render };

export { Accordion };
export default meta;
