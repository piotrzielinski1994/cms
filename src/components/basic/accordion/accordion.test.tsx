import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { Accordion } from './accordion';

describe('Accordion', () => {
  const defaultItems = [
    { heading: 'Heading 1', content: 'Content 1' },
    { heading: 'Heading 2', content: 'Content 2' },
  ] satisfies ComponentProps<typeof Accordion>['items'];

  it('should have no accessibility violations', async () => {
    const { container } = render(<Accordion items={defaultItems} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<Accordion items={defaultItems} />);
    expect(container).toMatchSnapshot();
  });
});
