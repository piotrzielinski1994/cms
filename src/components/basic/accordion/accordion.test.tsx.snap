// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > should match the snapshot 1`] = `
<div>
  <div
    class="grid gap-2"
  >
    <div
      class="bg-components-accordion text-components-accordion-foreground"
    >
      <button
        aria-controls="«r1»__0"
        aria-expanded="false"
        class="w-full px-4 py-2 flex justify-between items-center gap-2 text-lg"
        type="button"
      >
        <span>
          Heading 1
        </span>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down h-[1em] w-[1em] transition-transform"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <div
        class="overflow-hidden transition-[max-height] duration-300 ease-in-out"
        id="«r1»__0"
        style="max-height: 0;"
      >
        <div
          class="px-4 py-2"
        >
          Content 1
        </div>
      </div>
    </div>
    <div
      class="bg-components-accordion text-components-accordion-foreground"
    >
      <button
        aria-controls="«r1»__1"
        aria-expanded="false"
        class="w-full px-4 py-2 flex justify-between items-center gap-2 text-lg"
        type="button"
      >
        <span>
          Heading 2
        </span>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down h-[1em] w-[1em] transition-transform"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
      <div
        class="overflow-hidden transition-[max-height] duration-300 ease-in-out"
        id="«r1»__1"
        style="max-height: 0;"
      >
        <div
          class="px-4 py-2"
        >
          Content 2
        </div>
      </div>
    </div>
  </div>
</div>
`;
